walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // If request duration is longer than a day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If no one has other events
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> mandatoryBusyTimes = new ArrayList<TimeRange>();
+    List<TimeRange> optionalBusyTimes = new ArrayList<TimeRange>();
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    
+    // Get mandatory attendees' busy times
+    if (request.getAttendees().size() > 0) {
+      mandatoryBusyTimes = getBusyTimes(events, request.getAttendees());
+      busyTimes.addAll(mandatoryBusyTimes);
+    }
+
+    // Get optional attendees' busy times
+    if (request.getOptionalAttendees().size() > 0) {
+      optionalBusyTimes = getBusyTimes(events, request.getOptionalAttendees());
+      busyTimes.addAll(optionalBusyTimes);
+    }
+    
+    // If there are busy times, find the open times that can fit the meeting duration
+    if (busyTimes.size() > 0) {
+      Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+      busyTimes = mergeOverlappedTimes(busyTimes);
+      List<TimeRange> openTimes = getOpenTimes(busyTimes, request.getDuration());
+      
+      // If there are no open times with the optional + mandatory, find open times for only the mandatory 
+      if (openTimes.size() > 0) {
+        return openTimes;
+      }
+      if (request.getAttendees().size() > 0) {
+        Collections.sort(mandatoryBusyTimes, TimeRange.ORDER_BY_START);
+        return getOpenTimes(mandatoryBusyTimes, request.getDuration());
+      }
+      // If there are no open times with the optional + mandatory and no mandatory attendees for the event,
+      return Arrays.asList();
+    }
+
+    // Return the whole day if there are no busy times
+    return Arrays.asList(TimeRange.WHOLE_DAY);
```

Comment:
I think we normally do early return for edge cases?

Reply:
Oh yeah I'm not sure why I had it like that. Fixed it!